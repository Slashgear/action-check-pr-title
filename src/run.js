const core = require('@actions/core');
const github = require('@actions/github');


exports.run = () => {
  const { eventName } = github.context;
  core.info(`Event name: ${eventName}`);

  if (eventName !== 'pull_request') {
    core.setFailed(`Invalid event: ${eventName}, it should be use on pull_request`);
    return;
  }

  const pullRequestTitle = github.context.payload.pull_request.title;

  core.info(`Pull Request title: "${pullRequestTitle}"`);

  const regex = RegExp(core.getInput('regexp'));
  if (!regex.test(pullRequestTitle)) {
    core.setFailed(`Pull Request title "${pullRequestTitle}" failed to pass match regex - ${regex}`);
  }
};
