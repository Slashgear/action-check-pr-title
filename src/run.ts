import { info, setFailed, getInput } from "@actions/core";
import { Context } from "@actions/github/lib/context";

export const run = (context: Context) => {
  const { eventName } = context;
  info(`Event name: ${eventName}`);

  if (eventName !== "pull_request") {
    setFailed(`Invalid event: ${eventName}, it should be use on pull_request`);
    return;
  }

  const pullRequestTitle = context?.payload?.pull_request?.title;

  info(`Pull Request title: "${pullRequestTitle}"`);

  const regex = RegExp(getInput("regexp"), getInput("flags"));
  const helpMessage = getInput("helpMessage");
  if (!regex.test(pullRequestTitle)) {
    let message = `Pull Request title "${pullRequestTitle}" failed to pass match regexp - ${regex}
`;
    if (helpMessage) {
      message = message.concat(helpMessage);
    }

    setFailed(message);
  }
};
