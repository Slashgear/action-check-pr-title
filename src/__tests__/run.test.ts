// oxlint-disable consistent-type-imports
import core from "@actions/core";
import type { Context } from "@actions/github/lib/context";
import { Mock, beforeEach, describe, expect, it, vi } from "vitest";
import { run } from "../run";

vi.unmock("../run");
// @ts-expect-error
vi.mock<typeof import("@actions/github")>("@actions/github", () => ({
  context: {
    eventName: "pull_request",
    payload: {
      pull_request: {
        title: "This is a pull request title",
      },
    },
  },
}));

vi.fn();

const mockInputValues = (vitestFn: Mock, mocks: unknown) => {
  vitestFn.mockImplementation((input: unknown) => {
    // @ts-ignore
      return { regexp: "", flags: "", helpMessage: "", ...mocks }[input];
  });
};

describe("run", () => {
  beforeEach(() => {
    vi.spyOn(core, "getInput");
    vi.spyOn(core, "setFailed");
  });

  it("should pass nicely if title match regexp", () => {
    mockInputValues(core.getInput as Mock, { regexp: "^[a-z ]+$", flags: "i" });
    run({
      eventName: "pull_request",
      payload: {
        pull_request: {
          title: "This is a pull request title",
        },
      },
    } as unknown as Context);
    expect(core.setFailed).not.toHaveBeenCalled();
  });

  describe("on failing", () => {
    let regexp: string;
    let pullRequestTitle;
    let context: Context;

    beforeEach(() => {
      regexp = String.raw`\d`;
      pullRequestTitle = "This is a pull request title";
      context = {
        eventName: "pull_request",
        payload: {
          pull_request: {
            number: 1,
            title: pullRequestTitle,
          },
        },
      } as unknown as Context;
    });

    it("should fails on regexp matching", () => {
      mockInputValues(core.getInput as Mock, { regexp });
      run(context);
      expect((core.setFailed as Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it("should fails on regexp matching with helper message if defined", () => {
      mockInputValues(core.getInput as Mock, {
        regexp,
        helpMessage: `Example of matching titles:
"[Example] example of title (US-6596)"
`,
      });
      run(context);
      expect((core.setFailed as Mock).mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
