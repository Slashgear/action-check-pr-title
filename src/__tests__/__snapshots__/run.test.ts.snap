// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`run > on failing > should fails on regexp matching 1`] = `
"Pull Request title "This is a pull request title" failed to pass match regexp - /\\d/
"
`;

exports[`run > on failing > should fails on regexp matching with helper message if defined 1`] = `
"Pull Request title "This is a pull request title" failed to pass match regexp - /\\d/
Example of matching titles:
"[Example] example of title (US-6596)"
"
`;
